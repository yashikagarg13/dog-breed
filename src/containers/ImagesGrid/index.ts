export { default } from './ImagesGrid'
export type { ImagesGridProps } from './ImagesGrid'