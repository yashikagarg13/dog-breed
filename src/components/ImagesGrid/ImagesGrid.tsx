import React, { useEffect, useState } from 'react'
import { makeStyles } from '@material-ui/core/styles'
import CircularProgress from '@material-ui/core/CircularProgress'
import { ImagesGridProps } from './../../containers/ImagesGrid'
import { Images } from './../../reducers/images'

const useStyles = makeStyles(() => ({
  imagesContainer: {
    marginTop: 30,
    marginBottom: 30
  },
  error: {
    color: 'red'
  },
  imagesTitle: {
    textTransform: 'capitalize',
  },
  imageGrid: {
    display: 'flex',
    gridAutoColumns: 150,
    gap: 15,
    flexWrap: 'wrap'
  },
  image:{
    width: 150,
    height: 150
  }
}))
function ImagesGrid(props: ImagesGridProps) {
  const classes = useStyles()
  const { breeds, images, error, loading, fetchImages } = props
  const [imagesForSelectedBreeds, setImagesForSelectedBreeds] = useState<Images>({})

  useEffect(() => {
    breeds.length && fetchImages(breeds)
  }, [breeds, fetchImages])

  useEffect(() => {
    const selectedImages = breeds.reduce((filteredImages: Images, breed) => {
      filteredImages[breed] = images[breed]
      return filteredImages
    }, {})
    setImagesForSelectedBreeds(selectedImages)
  }, [images, breeds])

  return (
    <div className={classes.imagesContainer}>
      {loading ? (
          <CircularProgress />
        ) : (
          <>
            {error && <small className={classes.error}>{error}</small>}
            {imagesForSelectedBreeds && (
              Object.keys(imagesForSelectedBreeds).map((breed: string) => (
                <div key={breed}>
                  <h4 className={classes.imagesTitle}>{breed}</h4>
                  <div className={classes.imageGrid}>
                    {(imagesForSelectedBreeds[breed] || []).map((image: string, index: number) => (
                      <img 
                        key={index} 
                        className={classes.image} 
                        src={image} 
                        alt={`${breed}-${index + 1}`} 
                      />
                    ))}
                  </div>
                </div>
              ))
            )}
          </>
        )
      }
    </div>
  )
}

export default ImagesGrid