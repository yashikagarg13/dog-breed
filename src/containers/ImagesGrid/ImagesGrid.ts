import { connect, ConnectedProps } from 'react-redux'
import {
  getImages,
  getImagesError,
  isImagesLoading,
  isImagesFetched
} from '../../selectors'
import { AppDispatch, RootState } from '../../store'
import { fetchImages } from '../../reducers/images'
import ImagesGrid from '../../components/ImagesGrid'

const mapStateToProps = (state: RootState) => ({
  images: getImages(state),
  error: getImagesError(state),
  loading: isImagesLoading(state),
  fetched: isImagesFetched(state)
})

const mapDispatchToProps = (dispatch: AppDispatch) => ({
  fetchImages: (breeds: string[]) => dispatch(fetchImages(breeds))
})

const connector = connect(
  mapStateToProps,
  mapDispatchToProps
)

type ImagesGridPropsRedux = ConnectedProps<typeof connector>

export type ImagesGridProps = ImagesGridPropsRedux & {
  breeds: string[]
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(ImagesGrid) 
