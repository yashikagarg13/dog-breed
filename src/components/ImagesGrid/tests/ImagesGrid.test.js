import React from 'react'
import { render, fireEvent } from '@testing-library/react'

import ImagesGrid from '../ImagesGrid'

describe('ImagesGrid Component', () => {
  let props
  const breeds = ['hound', 'hound: test']
  const images = {
    'hound': [
      'test-image-1',
      'test-image-2',
      'test-image-3'
    ],
    'hound: test': [
      'test-image-4',
      'test-image-5'
    ]
  }

  beforeEach(() => {
    props = {
      breeds, 
      images, 
      error: '', 
      loading: false, 
      fetchImages: jest.fn()
    }
  })

  it('should render ImagesGrid component', () => {
    const { container } = render(
      <ImagesGrid {...props} />
    )
    expect(container).toBeInTheDocument()
    expect(container.innerHTML).toMatchSnapshot()
  })

  it('should call fetchImages if breeds changes', () => {
    props.breeds = ['new-breed']
    const { container } = render(
      <ImagesGrid {...props} />
    )

    expect(props.fetchImages).toHaveBeenCalledWith(props.breeds)
  })

  it('should not call fetchImages if breeds changes', () => {
    props.breeds = []
    const { container } = render(
      <ImagesGrid {...props} />
    )

    expect(props.fetchImages).not.toHaveBeenCalled()
  })

  it('should show loader if loading is true', () => {
    props.loading = true
    const { container } = render(
      <ImagesGrid {...props} />
    )
    expect(container).toBeInTheDocument()
    expect(container.innerHTML).toMatchSnapshot()
  })


  it('should show error message if error message is set', () => {
    props.error = 'error'
    const { container } = render(
      <ImagesGrid {...props} />
    )
    expect(container).toBeInTheDocument()
    expect(container.innerHTML).toMatchSnapshot()
  })
})
