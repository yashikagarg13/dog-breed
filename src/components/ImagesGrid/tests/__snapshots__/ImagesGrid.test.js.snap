// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImagesGrid Component should render ImagesGrid component 1`] = `"<div class=\\"makeStyles-imagesContainer-1\\"><div><h4 class=\\"makeStyles-imagesTitle-3\\">hound</h4><div class=\\"makeStyles-imageGrid-4\\"><img class=\\"makeStyles-image-5\\" src=\\"test-image-1\\" alt=\\"hound-1\\"><img class=\\"makeStyles-image-5\\" src=\\"test-image-2\\" alt=\\"hound-2\\"><img class=\\"makeStyles-image-5\\" src=\\"test-image-3\\" alt=\\"hound-3\\"></div></div><div><h4 class=\\"makeStyles-imagesTitle-3\\">hound: test</h4><div class=\\"makeStyles-imageGrid-4\\"><img class=\\"makeStyles-image-5\\" src=\\"test-image-4\\" alt=\\"hound: test-1\\"><img class=\\"makeStyles-image-5\\" src=\\"test-image-5\\" alt=\\"hound: test-2\\"></div></div></div>"`;

exports[`ImagesGrid Component should show error message if error message is set 1`] = `"<div class=\\"makeStyles-imagesContainer-21\\"><small class=\\"makeStyles-error-22\\">error</small><div><h4 class=\\"makeStyles-imagesTitle-23\\">hound</h4><div class=\\"makeStyles-imageGrid-24\\"><img class=\\"makeStyles-image-25\\" src=\\"test-image-1\\" alt=\\"hound-1\\"><img class=\\"makeStyles-image-25\\" src=\\"test-image-2\\" alt=\\"hound-2\\"><img class=\\"makeStyles-image-25\\" src=\\"test-image-3\\" alt=\\"hound-3\\"></div></div><div><h4 class=\\"makeStyles-imagesTitle-23\\">hound: test</h4><div class=\\"makeStyles-imageGrid-24\\"><img class=\\"makeStyles-image-25\\" src=\\"test-image-4\\" alt=\\"hound: test-1\\"><img class=\\"makeStyles-image-25\\" src=\\"test-image-5\\" alt=\\"hound: test-2\\"></div></div></div>"`;

exports[`ImagesGrid Component should show loader if loading is true 1`] = `"<div class=\\"makeStyles-imagesContainer-16\\"><div class=\\"MuiCircularProgress-root MuiCircularProgress-colorPrimary MuiCircularProgress-indeterminate\\" style=\\"width: 40px; height: 40px;\\" role=\\"progressbar\\"><svg class=\\"MuiCircularProgress-svg\\" viewBox=\\"22 22 44 44\\"><circle class=\\"MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate\\" cx=\\"44\\" cy=\\"44\\" r=\\"20.2\\" fill=\\"none\\" stroke-width=\\"3.6\\"></circle></svg></div></div>"`;
